Added RPC calls for MsigCreate, MsigPropose, MsigApprove, and MsigCancel

aagdere opened the pull request:
I added the RPC commands that correspond to the cli commands and removed duplicate code so the cli refers to the RPC API. 

- `lotus send` -> `Filecoin.Send`
- `lotus msig create` -> `Filecoin.MsigCreate`
- `lotus msig propose` -> `Filecoin.MsigPropose`
- `lotus msig approve` -> `Filecoin.MsigApprove`

The cli behavior should not be affected by this, purely a structural change.

Below I attached two bash scripts that I used to test the changes. [MsigSendExample.sh.txt](https://github.com/filecoin-project/lotus/files/4576999/MsigSendExample.sh.txt) is for cli commands to make sure this change is backwards compatible, and 
[MsigRPCSendExample.sh.txt](https://github.com/filecoin-project/lotus/files/4577013/MsigRPCSendExample.sh.txt) replaces some of the cli commands to test the new RPC calls.

New RPC example calls:

Send:
```
curl -X POST -H "Content-Type: application/json" \
-H "Authorization: Bearer $(cat ~/.lotus/token)" \
--data '{"jsonrpc": "2.0","method":"Filecoin.Send",'\
'"params": ['\
'"t1by5koc5um5hz4rbuerxqf3y6uqp3s7n4txxkhoi", '\
'"1000000000000000000", '\
'"t3viyi4icxrwblk6kuvi4i3lq7x5khedhqrn6iy3x4oqxyzgpday3qca4q5kdxw3l3vkwgqxn2tpxgefsmqbqa", '\
'"0"], '\
'"id":3}' http://0.0.0.0:1511/rpc/v0 | jq
```
Output:
```
{
  "jsonrpc": "2.0",
  "result": {
    "/": "bafy2bzacebw2mou2ks2gpo6yss3fqu3iqmlpqn73k7nnx6hmcjfcseyh4f6m4"
  },
  "id": 3
}
```

MsigCreate:
```
curl -X POST \
-H "Content-Type: application/json" \
-H "Authorization: Bearer $(cat ~/.lotus/token)" \
--data '{"jsonrpc": "2.0","method":"Filecoin.MsigCreate",'\
'"params": ['\
'2, '\
'["t01001", "t01002", "t01003"], '\
'"1000000000000000000", '\
'"t3viyi4icxrwblk6kuvi4i3lq7x5khedhqrn6iy3x4oqxyzgpday3qca4q5kdxw3l3vkwgqxn2tpxgefsmqbqa", '\
'"0"], '\
'"id":3}' http://0.0.0.0:1511/rpc/v0 | jq
```
Output:
```
{
  "jsonrpc": "2.0",
  "result": {
    "IDAddress": "t01004",
    "RobustAddress": "t2ljbq5usn6wp3m2veyh7qdehzp7hp2hzex52ledi"
  },
  "id": 3
}
```
MsigPropose:
```
curl -X POST \
-H "Content-Type: application/json" \
-H "Authorization: Bearer $(cat ~/.lotus/token)" \
--data '{"jsonrpc": "2.0","method":"Filecoin.MsigPropose",'\
'"params": ['\
'"t23pwta2z6oezwpsr2glcrqnzuhlilqt2cbfgygmy", '\
'"t1gvza63hsouazauudnq4r7xg44dp6rwymd5ntklq", '\
'"1000000000000000000", '\
'"t01006", '\
'0, '\
'""],'\
' "id":3}' http://0.0.0.0:1511/rpc/v0 | jq
```
Output:
```
{
  "jsonrpc": "2.0",
  "result": {
    "MessageCid": {
      "/": "bafy2bzacediffrqzwsk3oxrga66haitib2jf75z24jcukiss2ejwxnragai2g"
    },
    "TransactionId": 1
  },
  "id": 3
}
```
MsigApprove:
```
curl -X POST \
-H "Content-Type: application/json" \
-H "Authorization: Bearer $(cat ~/.lotus/token)" \
--data '{"jsonrpc": "2.0","method":"Filecoin.MsigApprove",'\
'"params": ['\
'"t23shtwmzzdwo3harzxyyw2noaxppcwvjeptz6ypq", '\
'0, '\
'"t01006", '\
'"t1gvza63hsouazauudnq4r7xg44dp6rwymd5ntklq", '\
'"1000000000000000000", '\
'"t01007", '\
'0, '\
'""], '\
'"id":3}' http://0.0.0.0:1511/rpc/v0 | jq
```
Output:
```
{
  "jsonrpc": "2.0",
  "result": {
    "/": "bafy2bzacedatacqeoicpsh7jpzi76v2gzupqor5gohfkjqlibxp72gupisr3w"
  },
  "id": 3
}
```

whyrusleeping commented:
Hrm... we've been leaning away from making more RPC endpoints when you can get the job done with combinations of existing ones. Maybe multisig operations should be special?

Would we also want RPC endpoints for key changes? cancellations? hard questions

cc @magik6k @Kubuxu

whyrusleeping commented:
(in any case, thanks for the PR, and the well written explanation!)

magik6k commented on api/api_full.go:
```diff
@@ -163,6 +166,8 @@ type FullNode interface {
 	PaychVoucherList(context.Context, address.Address) ([]*paych.SignedVoucher, error)
 	PaychVoucherSubmit(context.Context, address.Address, *paych.SignedVoucher) (cid.Cid, error)
+
+	Send(context.Context, address.Address, types.BigInt, address.Address, types.BigInt) (cid.Cid, error)
```
I'm not sure if this methods does enough to be here, it's basically just an alias to MpoolPushMessage which optionally uses default wallet address.

I mean `Send` is kind of important, and it's a much more obvious name than `MpoolPushMessage`, but I'd prefer renaming `MpoolPushMessage` to something else, rather than have 2 mostly identical methods

aagdere replied:
That's a good point. I'll separate it out into another PR just for now just to get this merged in faster, but I'd also understand closing it out entirely. 

Like you mentioned, the logic for Send was just to have a go-to command for people that don't know the significance of things like gas or nonce. But they can just leave those out of the message they supply right? Not too familiar with Go.

magik6k commented on api/api_full.go:
```diff
@@ -147,6 +147,9 @@ type FullNode interface {
 
 	MsigGetAvailableBalance(context.Context, address.Address, types.TipSetKey) (types.BigInt, error)
+	MsigCreate(context.Context, int64, []address.Address, types.BigInt, address.Address, types.BigInt) (AddressPair, error)
+	MsigPropose(context.Context, address.Address, address.Address, types.BigInt, address.Address, uint64, string) (MsigProposeOutput, error)
```
Params should just be raw `[]byte`s, not hex encoded string

aagdere replied:
Changed `params` from `string` to `[]byte` in both `MsigPropose` and `MsigApprove`

droark commented on chain/types/bigint.go:
```diff
@@ -63,6 +63,10 @@ func BigCmp(a, b BigInt) int {
 	return a.Int.Cmp(b.Int)
 }
 
+func BigPos(bi BigInt) bool {
```
I think [`Sign()`](https://golang.org/pkg/math/big/) might make this redundant.

aagdere replied:
Sorry I took a look at the link but it wasn't directly obvious how that would help. Will it tell me if a given int is positive?

droark replied:
[Does this help?](https://golang.org/pkg/math/big/#Int.Sign) I think it'll work but I haven't tried it. Barring that, after looking at the source code a little more closely, I think you can just use `bi.neg` would give you a boolean indicating if it's negative. If neither one of them works, then yeah, your code's probably as good as it gets.

whyrusleeping replied:
Yeah, just use `.Sign()` for this. if it returns a number greater than zero, its positive. If it returns zero, its zero, if it returns a negative number, its negative

droark commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,271 @@
+	}, nil
+}
+
+func (a *MsigAPI) MsigApprove(ctx context.Context, msig address.Address, txId uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params string) (cid.Cid, error) {
```
Adding a `Cancel` method would be great for the sake of completeness. AFAIK, it's just `Approve` with a different method code.

aagdere replied:
I'm going to need to look into how to Cancel a proposal. Maybe you can help me with that if you know how?

droark commented on cli/multisig.go:
```diff
@@ -106,69 +100,21 @@ var msigCreateCmd = &cli.Command{
-			return err
-		}
-		fmt.Println("Created new multisig: ", execreturn.IDAddress, execreturn.RobustAddress)
+		fmt.Println("Created new multisig: ", addressPair.IDAddress, addressPair.RobustAddress)
```
Somebody more familiar with Lotus can speak to this but I'm not sure if exposing the robust address is a good idea. (This is the first I've heard of it, honestly!) If nothing else, they should be marked as ID and robust addresses.

aagdere replied:
To my knowledge there is no reason to keep this private. Also, this merely mimics the output of the cli command. `lotus msig create` currently returns both the ID and Robust addresses.

droark replied:
Fair enough. I am curious about `RobustAddress`, though. I haven't seen that before. Can you give some examples from your testing? Are they `t0` addresses or something else?

aagdere replied:
Yea sure! Take a look at the PR description above under `Output` for `MsigCreate`. Although I'm not sure if it is always guaranteed to be the case, in my example it is a t2 address (Actor Address) I'm assuming Robust just means non-ID address but @whyrusleeping or @magik6k can speak to that.

droark replied:
Thanks. Sounds like it's just another name for actor addresses but somebody else can chime in and confirm. :)

whyrusleeping replied:
Yeah, the robust address is stable across chain reorgs. The ID address may change if the block including your creation message is reorged

droark commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,271 @@
+	}
+
+	// wait for it to get mined into a block
+	wait, err := a.StateAPI.StateWaitMsg(ctx, smsg.Cid())
```
In case there's a holdup, adding a statement ("Waiting for the message to be mined") wouldn't hurt.

aagdere replied:
That's a good call. Especially since I included one in `MsigProposal`. 
Added `fmt.Println("send proposal in message: ", smsg.Cid())` to make this more symmetrical to the other one.

aagdere replied:
Added, resolving

droark commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,271 @@
+		return api.UndefMsigProposalOutput, actErr
+	}
+
+	msg := &types.Message{
```
There might be some DRYing opportunities here.

aagdere replied:
Hmm how would you recommend I go about that? The only DRY I really practiced in this PR was moving the cli logic into the RPC calls.

droark replied:
I don't have any immediate suggestions for DRYing. In cases like these, I just look at the code and try to determine if there are any patterns, like two or more calls making, say, three or more calls that are exactly the same or only vary in one parameter or so. It's a little tricky. There's a fine line between DRYing that makes sense and DRYing that can potentially hamper maintenance due to creating functions that make little logical sense. If you don't see anything, don't sweat it. :)

aagdere replied:
I just spent some time comparing the three functions. The three functions seem too different to merge logic. If anything I was hoping to find some similarities between Propose and Approve, but most of the logic is input validation and serialization which differs since they take different parameters :/ I'm thinking if I add a `Cancel` function, it should be easy to factor out some common code with `Propose` since they should take the same params. I'll start to tackle that one now.

droark replied:
Fair enough. Nothing immediately jumped out at me after looking a little more closely but I thought there might be a chance. That said, yes, I think `Approve`/`Cancel` should have more opportunities since they're essentially twins, at least in terms of what goes out on the wire.

aagdere commented:
Addressed some of your comments, should be ready for another round of review. 

@droark still need to get back to you on the Cancel method. Until now I wasn't even aware that you can cancel a proposal. I'll have to find out how before I can implement the RPC command.

droark commented:
> @droark still need to get back to you on the Cancel method. Until now I wasn't even aware that you can cancel a proposal. I'll have to find out how before I can implement the RPC command.

Cool. AFAIK, it really is exactly the same as `Approve` in terms of structure, only you use a different method code. `builtin.MethodsMultisig.Cancel` should do the trick.

aagdere commented:
Added `MsigCancel` and fixed the `golangci-lint` issues. Was going to fix the goimports issue as well but that caused merge conflicts (I'm guessing you guys already took care of that). 

Anyway, this build should go totally green if the `goimports` issue is fixed. @whyrusleeping @magik6k I made your suggested changes. Let me know if it looks good to merge or if you'd like me to change anything else. I can't spend too much more time on this but this branch is enough to unblock me for now so I'll be using it to build my local devnet node.

droark commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,282 @@
+	return a.MsigApproveOrCancel(ctx, api.MsigCancel, msig, txID, proposer, to, amt, src, method, params)
+}
+
+func (a *MsigAPI) MsigApproveOrCancel(ctx context.Context, operation api.MsigProposeResponse, msig address.Address, txID uint64, proposer address.Address, to address.Address, amt types.BigInt, src address.Address, method uint64, params []byte) (cid.Cid, error) {
```
This should be private. Assuming I understand Go correctly, changing the name to `msigApproveOrCancel` should do the trick.

aagdere replied:
Was not aware of private methods in Go lol... Good call.

droark commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,282 @@
+		We pass in a MsigProposeResponse instead of MethodNum to
+		tighten the possible inputs to just Approve and Cancel.
+	*/
+	switch operation {
```
I'd move this up top, and introduce some sort of error condition if the function gets the wrong operation. As an alternative, I'd pass in a bool and set the method based on the bool. This way, you avoid a possible error condition.

aagdere replied:
I actually specifically opted for an enum over a bool to avoid such confusion. I think a boolean is less explicit than case matching. I'm not too familiar with Go so I'm not sure if it will be an exhaustive case match, but it shouldn't be possible to hit any weird state if you require the param to be a `MsigProposeResponse` unless you add a new case (which would be very deliberate). The null case however might be a problem, so it may be worth adding that case if it's a possibility.

droark replied:
Gotcha. I'm not necessarily opposed to a boolean, for the reasons you mentioned. The switch just needs to cover all cases, and either add an error condition or throw if passed the wrong method. I prefer the former, even with the slight addition of complexity.

droark commented:
@aagdere - Not sure if you saw it but apparently there's a conflict with `api/apistruct/struct.go`.

aagdere commented:
@droark Yea I saw that. That's kind of the annoying thing about the multiple commit approach as opposed to rebasing. If you merge and then make commits on top, rebasing becomes a headache and requires cherry-picking. I checked out the conflict, it's actually not even overlapping so I'm not sure why git is complaining.

Anyway, I'll just fix that up after the PR is approved during my commit squashing :)

droark commented:
> @droark Yea I saw that. That's kind of the annoying thing about the multiple commit approach as opposed to rebasing. If you merge and then make commits on top, rebasing becomes a headache and requires cherry-picking. I checked out the conflict, it's actually not even overlapping so I'm not sure why git is complaining.
> 
> Anyway, I'll just fix that up after the PR is approved during my commit squashing :)

Heh. Welcome to my world. Got to take a couple of lumps while working on that `specs-actors` PR. :)

whyrusleeping commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,282 @@
+		req = lenAddrs
+	}
+
+	if src == address.Undef {
```
I would do this logic on the client side, and require callers of the API to always pass an address here.

whyrusleeping commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,282 @@
+		src = defaddr
+	}
+
+	if gp == types.EmptyInt {
```
this is fine, can use an oracle for gas prices in the future

whyrusleeping commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,282 @@
+		return api.UndefMsigProposalOutput, xerrors.Errorf("must provide a positive amount for proposed send")
+	}
+
+	if src == address.Undef {
```
same comment here, force the caller to resolve this

whyrusleeping commented on node/impl/full/multisig.go:
```diff
@@ -0,0 +1,282 @@
+		return api.UndefMsigProposalOutput, nil
+	}
+
+	fmt.Println("send proposal in message: ", smsg.Cid())
```
avoid fmt.Printlns in api side calls, if we really need to log something (which i dont think is necessary here) use `log.{Warn,Info,Debug}`

aagdere commented:
Fixed Merge Conflicts
